src/transaction.rs

```diff
@@ -377,19 +417,23 @@ mod tests {
         let keypair1 = Keypair::new();
         let zero = Hash::default();
         let mut tx = Transaction::new(&keypair0, keypair1.pubkey(), 1, zero);
-        if let Instruction::NewContract(contract) = &mut tx.instruction {
+        let mut instruction = tx.instruction();
+        if let Instruction::NewContract(ref mut contract) = instruction {
             if let Plan::Budget(Budget::Pay(ref mut payment)) = contract.plan {
                 payment.tokens = 2; // <-- attack!
             }
         }
+        tx.userdata = serialize(&instruction).unwrap();
```

Comment:
Instead, consider calling `serialize(&instruction)` from `Transaction::new` or creating a new `Transaction::new_from_serializable`.

Reply:
@garious the original code modified the instruction in place, so seems like this is appropriate.

Reply:
The original code modifies transactions in place strictly to demonstrate attacks.

Reply:
@garious i thought that was the point.  semantically this is the same as in-place modification.  The constructed transaction fails the `verify_plan` call.  otherwise `verify_plan` should sbe a method on `Instruction`.  If that makes more sense, lets sync up to figure out how to organize the Instruction/budget dsl code separate from the transaction processing plumbing.

Reply:
Yes, you're right. Now I grok it. I thought you were maintaining the Instruction in both the `instruction` and `userdata` fields. Good stuff.